### ISSUE TITLE
Backport candidates for potential 2.2

### ISSUE THREAD

**Issue by keith-turner** [apache/accumulo#3092]
This issue exists to track potential 3.0 changes that may motivate a 2.2 release.  However there may never be a 2.2 release if there are not enough issues.  See this [discussion](https://github.com/apache/accumulo/pull/3088#discussion_r1027114511)

 * #3088

**Comment by EdColeman** [comment]
If this continues to grow, my opinion is that it should be a project.

Continuing the discussion - I very much prefer leveraging GH issues and projects for tracking things over a personnel lists  - that allows  everyone to track / have visibility into things and relieves me of the burden of making my own list, keeping it up to date and tracking it myself.

The advantage of projects is that is tracks multiple issues and shows outstanding vs completed independent of the individual issues.

Being able to review open issues on GH allows me to judge where things stand and what people are thinking about.  If there are areas of code that I think could use some attention, knowing what other potential changes should be considered might help deconflict what I might propose.  If it was aon someone's private list, then I have no visibility and need to rely on someone noticing my change and realizing that it may brush on things that could be impacted.

**Comment by ctubbsii** [comment]
Regardless of where you track this list (and I would prefer a "project" (we've been using the "classic" projects) over an issue that is more easily lost, I don't think we need to include things that are already staged for a 2.1 patch release, unless we're trying to argue that the 2.2 should replace the 2.1 as a continuation of that "LTM" series. But, that could get really confusing, suggesting that risk-averse users jump a minor release when they chose to stay on LTM releases to avoid the risks associated with new features.

Then again, we can just say "oops, our mistake", 2.1.0 is not an LTM release, but 2.2.0 is. But then, when does it stop? Maybe next time we do a .0 release, we do an initial release as non-LTM, and then only decide it's going to be an LTM release later? I don't know what's least confusing to users trying to track the version hell.

**Comment by ctubbsii** [comment]
I didn't see the earlier discussion on #3088 that suggested an issue instead of a project, so it's less noisy. Either way works for me.

**Comment by keith-turner** [comment]
@ctubbsii and @EdColeman I agree that GH projects are a better way to track this type of information.  I was concerned about it being noisy and the lack of consensus on wether 2.2.0 was even a good goal.  I am not even sure 2.2.0 is a good goal.  Creating a GH project named 2.2.0 feels like this is something we have reached consensus about.  I suppose a simple solution to that is to name the project 'possible 2.2.0' or something that clearly expresses the uncertainty.

Thinking about consensus I realized a few things.

  * When a PR is following previously settled consensus (like using semver) then its great because no time needs to be spent on the PR discussing the previously settled consensus.
  * When a PR is deviating from previously settled consensus, then a lot of time can spent on discussions around those deviations.

Overall having settled consensus saves us lots of time.   When we are spending lots of time in multiple PRs discussing deviations from settled consensus, then that could be an indication that we need to establish some new consensus or that we should stop deviating inorder to save ourselves the hassle.

Ive had specific thoughts about 2.2.0, but I have not been able to communicate them very well.  I realized what I was thinking about for 2.2.0 is something we have actually done before.  Going back to Accumulo 1.8.0 the releases looked like 1.8.0->1.8.1->1.9.0->1.9.1->1.9.2->1.9.3->1.10.0->1.10.1.  The [release notes from 1.9.0](https://accumulo.apache.org/release/accumulo-1.9.0/) contain the following 

```
Apache Accumulo 1.9.0 is a minor release on the 1.x branch. This release would be considered a maintenance release on 1.8 branch except there are some API additions which resulted in a new minor release.
```

The [release notes for 1.10.0](https://accumulo.apache.org/release/accumulo-1.10.0/) contained the following

```
Apache Accumulo 1.10.0 is a continuation of the 1.x release line, and is essentially the next maintenance release of 1.8/1.9, following the 1.9.3 version with some small additional internal improvements. Earlier 1.x versions are now superseded by this maintenance release, and will no longer be maintained.
```

We could follow the same strategy for 2.x starting w/ 2.1.0.  I don't know what to call this strategy though, besides what we did for 1.8.0.   I feel like we tried to do at least the following with 1.8.0 forward.

 * Continually improved stability by fixing serious bugs
 * Followed a principal of least surprise with the minor updates.  Not only did we keep the APIs super stable in 1.8.0 and forward, but we kept lots of stuff outside of the APIs stable like scripts, properties, config files, etc.  We could have made more disruptive changes in 1.9.0 and 1.10.0, but chose not to do so.

We could follow the same strategy for 2.1.0 forward (have minor releases that have the least suprise) if we develop consensus around it and properly communicate it AND still consider it a LTM line.  Like we could say the current LTM 1.10.x line started with 1.8.0.

One thing I have been thinking about is urgency in making this decision.   I don't think there is any urgency from the release perspective that would force a decision.  3.0.0 will not be released soon and 2.1.1 would not be released until we find serious bugs that motivate the release.  Currently there are no known serious bugs in 2.1.0.  One driver for urgency around deciding about 2.2.0 would be if making that decision could help us work more efficiently.  If we are not going to do 2.(1+x).0 releases, then maybe  we should decided that all PRs that violate semver in the 2.1.x line should just be pushed to 3.0.0 or later so that we can save ourselves time from rehashing.

Sorry for the tome. Thinking about consensus around the project name led to this deeper line of thinking about consensus in general.

**Comment by ctubbsii** [comment]
@keith-turner thank you for laying out your thoughts. One thing to consider regarding the path of least surprise that we've done before, is that we did that before establishing the LTM release strategy, which adds an extra dimension to consider. One of the goals of that is to keep a more and more stable feature fixed release line, and not just a minimal, least surprise, series of minor release feature additions. It's easy to communicate that 2.1 is LTM, but much more confusing to users that the version range `[2.1,2.3)` is a single LTM "version". It also complicated potential upgrade paths and maintenance branches we develop (would there be a 2.1.2 after a 2.2.0? Users may expect one because it's in the LTM range).

So, I'm inclined to bump things to 3.0 that would not be appropriate for 2.1 patch releases because of semver.

**Comment by keith-turner** [comment]
> would there be a 2.1.2 after a 2.2.0?

No, not if following was done w/ 1.8.0 onward.  It was a linear progression.

> So, I'm inclined to bump things to 3.0 that would not be appropriate for 2.1 patch releases because of semver.

Another option to consider is that we accept exceptions to semver in the 2.1.x line as long as they are well documented in release notes and javadoc.  This was discussed on one or more of the PRs.

**Comment by dlmarion** [comment]
Re-reading the [LTM](https://accumulo.apache.org/contributor/versioning#LTM) page, it doesn't preclude a 2.2.0, 2.3.0, etc., it just says that as a non-LTM release any patches/bugfixes would be released in a subsequent minor version. This allows us to continue to release features on the 2.x line (even experimental ones). The downside here is that we end up with more branches to maintain. I feel like if we did some planning we could find consensus on a path forward.

**Comment by ctubbsii** [comment]
@dlmarion You're right. But it's the "more branches to maintain" that I'm hoping to avoid. It might not be so bad if the changes are very minimal, but it would basically mean releasing a new 2.2\+ version every time we release a 2.1 LTM patch if we want people on those versions to also have the patch. However, I'd rather focus on 3.0, and put the proposed additions there. I just don't see much urgency to include these proposed features as a 2.x update instead of just rolling them into 3.0.

**Comment by dlmarion** [comment]
> But it's the "more branches to maintain" that I'm hoping to avoid.

I agree, that's a concern.

> However, I'd rather focus on 3.0, and put the proposed additions there. I just don't see much urgency to include these proposed features as a 2.x update instead of just rolling them into 3.0.

I think the issue here is that:

  1. 2.1 was decided as a LTM release too early
  2. we don't have a timeline for 3.x

I would be on board with moving my features to 3.x if I knew more about its timeline. In fact, I would propose something like (dates are hypothetical):

  3.0.0 - removal of deprecated items only (release date January or February - as early as possible)
  3.1.0 - feature set A (release June 2023)
  3.2.0 - feature set B (release Dec 2023)
  LTM - when we're done and need to make breaking changes for 4.x

**Comment by ctubbsii** [comment]
> * 2.1 was decided as a LTM release too early

I don't agree with that. LTM releases are intended to be feature frozen up front so we can have a target to maintain as stable. I think the issues we're seeing is growing pains as we get used to that idea, and we're seeing a desire to add features here and there that weren't ready (or even known to be desired) until too late. From my perspective, it feels like the intent of LTM is being undermined by a regret of not having everything we wanted. But the constant flux of features without a predetermined stable version that we were intending to maintain is the *main* problem we were trying to fix by having LTM releases in the first place. We can still have those features... in the next version... but not in the LTM one... that's the whole point. New features can wait until the next version, and users can decide for themselves if they want to stay on the LTM release or move forward with new features and incur additional risk. We let the user decide that by communicating very clearly what the LTM is and isn't.

> * we don't have a timeline for 3.x

That's fair, but we don't have a timeline for any 2.2 release either. Any plan for 2.2 to include these features can just be a plan for 3.0 to include those same features. I'm not opposed to a 2.2 in principal. I just don't think we need to limit how much we take on in terms of maintenance branches to reduce confusion for users and to manage our limited contributor resources. In fact, I think it's a good idea, in general, to make the LTM release the last minor release before moving to a new major release, because it helps users who aren't ready to upgrade and risk API breakages have confidence that the version will be maintained for a while and they won't have to upgrade.

> 3.0.0 - removal of deprecated items only (release date January or February - as early as possible)
> 3.1.0 - feature set A (release June 2023)

Your proposed timeline seems reasonable. Originally, I had wanted a 3.0 in December. It might be attainable, but would be rough with people taking time off for holidays. January seems more likely. I had wanted to push to keep 3.0.0 really minimal, in terms of new features, so we could release it quickly after stripping out deprecations to clean up and set the stage for subsequent changes free of old baggage that limited our development. However, I did not anticipate small new features that were ready so quickly, prior to 3.0.0 being released. I think it would be okay to add a few small changes in 3.0.0, as long as they don't delay the release, and aren't disruptive to the main cleanup tasks. I think the changes proposed so far are likely in that category.
